src/thor/triplegbuilder.cc

```diff
@@ -481,6 +482,72 @@ void SetHeadings(TripLeg_Edge* trip_edge,
   }
 }
 
+// TODO: how to use controller? or we don't need it at all?
+/**
+ * Add landmarks in directed edge to trip edge.
+ * @param  reader      Graph reader to get access to the graph tile and edge info.
+ * @param  trip_edge   Trip path edge to add landmarks.
+ * @param  controller  ?
+ * @param  edge        Directed edge where the landmarks are stored.
+ * @param  shape       Trip shape.
+ * @param  begin_index ?
+ */
+void AddLandmarks(GraphReader& reader,
+                  TripLeg_Edge* trip_edge,
+                  const AttributesController& controller,
+                  const DirectedEdge* edge,
+                  const std::vector<PointLL>& shape,
+                  const uint32_t begin_index) {
+  // get access to edge info from the graph tile
+  graph_tile_ptr tile = nullptr;
+  reader.GetGraphTile(edge->endnode(), tile); // TODO: why endnode() works (like everywhere else) ? is
+                                              // the edge always on the same tile?
+  auto ei = tile->edgeinfo(edge);
+
+  // get landmarks from tagged values in the edge info
+  for (const auto& tag : ei.GetTags()) {
+    if (tag.first == baldr::TaggedValue::kLandmark) {
+      Landmark lan(tag.second);
+      PointLL landmark_point = {lan.lng, lan.lat};
+
+      // find the closed point on edge to the landmark
+      auto closest = landmark_point.ClosestPoint(shape, begin_index);
+      const double kLandmarkMaxDisCutoff =
+          50.; // the maximum distance allowed between landmark and its closest point on the edge
+      if (std::get<1>(closest) > kLandmarkMaxDisCutoff) {
+        continue;
+      }
```

Comment:
maybe leave this out for now and in the future could allow a request option to have a tighter threshold